addons/hw_l10n_eg_eta/controllers/main.py

```diff
@@ -0,0 +1,170 @@
+    def eta_sign(self, pin, secret, invoices):
+        """
+        Gets cades-bes binaries from the odoo server for each invoice, signs each of them individually using the token
+        then returns the signatures
+        @param pin: pin of the token
+        @param secret: secret shared with the main odoo instance
+        @param invoices: dictionary of invoices. Keys are invoices ids, value are the base64 encoded binaries to sign
+        """
+        if not self._is_secret_valid(secret):
+            return self.unauthorized_error()
+        session, error = self._get_session(pin)
+        if error:
+            return error
+        try:
+            cert = session.findObjects([(PyKCS11.CKA_CLASS, PyKCS11.CKO_CERTIFICATE)])[0]
+            cert_id = session.getAttributeValue(cert, [PyKCS11.CKA_ID])[0]
+            priv_key = session.findObjects([(PyKCS11.CKA_CLASS, PyKCS11.CKO_PRIVATE_KEY), (PyKCS11.CKA_ID, cert_id)])
+
+            invoice_dict = dict()
+            invoices = json.loads(invoices)
+            for invoice, eta_inv in invoices.items():
+                to_sign = base64.b64decode(eta_inv)
+                signed_data = session.sign(priv_key, to_sign, PyKCS11.Mechanism(PyKCS11.CKM_SHA256_RSA_PKCS, None))
+                invoice_dict[invoice] = base64.b64encode(signed_data).decode('utf-8')
+
+            payload = {
+                'invoices': json.dumps(invoice_dict),
+            }
+            return json.dumps(payload)
+        except Exception as ex:
+            return self.get_error_template(str(ex), '')
+        finally:
+            session.logout()
+            session.closeSession()
+
+    def _get_session(self, pin):
+        session = False
+
+        lib, error = self.get_crypto_lib()
+        if error:
+            return session, error
+
+        try:
+            pkcs11 = PyKCS11.PyKCS11Lib()
+            pkcs11.load(pkcs11dll_filename=lib)
+        except PyKCS11Error as ex:
+            return session, self.get_error_template(str(ex), '')
+
+        slots = pkcs11.getSlotList(tokenPresent=True)
+        if not slots:
+            return session, self.get_error_template('No drive found - Make sure the thumb drive is correctly inserted',
+                                            'تأكد من أن الذاكرة الخارجية مدخلة بشكل صحيح - لم يتم العثور على أي محرك ذاكرة ')
+        if len(slots) > 1:
+            return session, self.get_error_template(
+                'Multiple drive detected - Only one secure thumb drive can be inserted at the same time',
+                'يمكن إدراج ذاكرة خارجية آمنة واحدة فقط في آن واحد - تم رصد عدة محركات ذاكرة')
+
+        try:
+            session = pkcs11.openSession(slots[0], PyKCS11.CKF_SERIAL_SESSION | PyKCS11.CKF_RW_SESSION)
+            session.login(pin)
+        except Exception as ex:
+            error = self.get_error_template(str(ex), '')
+        return session, error
+
+    def request_get(self, url, payload):
```

Comment:
These request_get and post seem like relics from the past as we do not need server to server communication anymore?